vivisect/__init__.py

```diff
@@ -920,6 +920,12 @@ def detectString(self, va):
                     if loc[L_LTYPE] == LOC_STRING:
                         if loc[L_VA] == va:
                             return loc[L_SIZE]
+                        if ord(bytez[offset+count]) != 0:
```

Comment:
this covers the "find the big string after the little one was already discovered" case.
does this catch the "find the little string in the middle of an existing big string" case?
if so, how?

this is getting complex enough that we probably want some explanation of what's going on (specifically wrt string within a string) use-cases and how they're handled.

Reply:
oh, nice point on this only capture suffixes

Reply:
counter point: rather than building the string index into the location database, what about pulling it out into its own object.

```py
class StringIndex:
  def find(self, needle) -> Tuple[str, Loc]
```

Reply:
yeah, there's some complexity in persisting the database.

what if we rebuilt the database upon each load? if we keep string start locations in the .viv, then its an easy operation to iterate, extract the strings, and index in memory. upon close, throw out the index.

of course it takes some CPU and memory to keep an index in memory, but i'd argue that until you can prove otherwise, its likely in line with the overhead of storing all the substrings in the .viv.

Reply:
also, could build the index lazily on first access which would help the common case where substrings aren't accessed.

Reply:
I wouldn't be against rebuilding the database on every load. Wouldn't be the worst thing in the world.

If we do go that route, I can see two possible ways of doing things. First is pushing some logic into the add location handler in vivisect/base.py so that on a makeString/makeUnicode call it does something similar-ish to what I've got here, in that it handles what the base string is, indexes, and so forth. Mostly so when we load an existing workspace, we end up with the same info as when we directly analyzing a binary, since workspace loading only ever sees the event stream. Second is similar to the first, except we plumb a new event type that handles all that for us.

I like the idea of a dedicated string database, as I think it would give us (easier) finer grained control over deletion, since right now it's "delete the main string and the substrings go with it". It's a more matter of do it now or do it later.